### PR TITLE
ci: add slither via slither-action

### PR DESCRIPTION
**Pull request by mds1**
## Overview

Closes https://github.com/ethereum-optimism/security-pod/issues/67

Uses [slither-action](https://github.com/crytic/slither-action) to add Slither to CI using GitHub Actions. This PR is my proposed first pass at finding a sustainable approach to continuous slither scanning that is not too noisy, and we can always adjust or remove it if this doesn't work out.

As explained in the updated README:

> `contracts-bedrock` uses [slither](https://github.com/crytic/slither) as its primary static analysis tool. Slither will be run against PRs as part of CI, and new findings will be reported as a comment on the PR. CI will fail if there are any new findings of medium or higher severity, as configured in the repo's Settings > Code Security and Analysis > Code Scanning > Protection rules setting.
> There are two corresponding jobs in CI: one calls "Slither Analysis" and one called "Code scanning results / Slither". The former will always pass if Slither runs successfully, and the latter will fail if there are any new findings of medium or higher severity.
> 
> Existing findings can be found in the repo's Security tab > [Code Scanning](https://github.com/ethereum-optimism/optimism/security/code-scanning) section. You can view findings for a specific PR using the `pr:{number}` filter, such [`pr:9405`](https://github.com/ethereum-optimism/optimism/security/code-scanning?query=is:open+pr:9405).
> 
> For each finding, either fix it locally and push a new commit, or dismiss it through the PR comment's UI.
> 
> Note that you can run slither locally by running `slither .`, but because it does not contain the triaged results from GitHub, it will be noisy. Instead, you should run `slither ./path/to/contract.sol` to run it against a specific file.

Note that previously the repo setting was high or higher, but I've changed it to medium or higher as there can be non-trivial medium findings from slither. Here is an example of what findings look like on github:
image

I triaged away the initial findings and committed them as `slither.db.json` to avoid polluting the UI with historical findings we don't care about.

There are two untriaged findings that I'm unsure if we want to fix, which you can find [here](https://github.com/ethereum-optimism/optimism/security/code-scanning?query=is:open+pr:9405). I will add a comment on this PR for each existing finding so we can decide what to do with it.

You can see an example of a finding triaged from the UI [here](https://github.com/ethereum-optimism/optimism/security/code-scanning/38):
image

On this PR, both Slither CI jobs are expected to pass because no *new* findings were introduced.

### CODE REVIEW

File: .github/workflows/slither.yml
Comment (mds1) [inline .github/workflows/slither.yml:1]: Using this comment to block merge, seeing some unexpected slither behavior:
1. Running `slither .` locally (after triage and with config file) there are 3 findings
2. In CI, slither has 50 findings https://github.com/ethereum-optimism/optimism/actions/runs/7808753591/job/21299473436?pr=9405#step:5:6521
3. The code scanning job is listed as passing with no new findings even though there were 50 (https://github.com/ethereum-optimism/optimism/pull/9405/checks?check_run_id=21299634075) 

What I expected was for 3 findings to show up in the security tab of the repo + 3 comments on the PR. I suspect this setting is the culprit, and that we need to change the security setting to Medium or higher. But I'm hesitant to do this until (2) is fixed because I don't want to pollute repo with unexpected findings

The cause appears to be that it's not picking up my slither.db.json, because I get 50 findings when running locally without the DB, so presumably just a monorepo path issue and need to figure out where to set that

image

Comment (mds1) [reply .github/workflows/slither.yml:1]: Created https://github.com/crytic/slither-action/issues/70, switching to draft until that is resolved

**Comment by tynes** [comment]
Did you mean to commit in the slither db?

**Comment by mds1** [comment]
Yes I committed it intentionally so devs can now run slither without all the noise of prior findings, since they've been triaged and the DB tells slither to ignore them. Once this PR is fixed/merged and the above 3 items addressed, the flow will be that devs can run `pnpm slither` on their branch to see new findings introduced by their PR, since everything else was triaged. Still discussing in https://github.com/crytic/slither-action/issues/70#issuecomment-1932335521 to iron out the best flow/UX here, and feedback is definitely welcome

I formatted the JSON so that it's easier to review a diff of the JSON and see what changed. But the default DB output was one line of JSON, and we can switch back to that format if you prefer. Alternatively can update `.gitattributes` to have it ignore the DB file

Edit: Committed to using only the github code scanning approach for now, so JSON is back down to 1 line just to prune historical outputs. Slither will soon implement a feature that only reports findings based on the diff between the current branch and a baes branch, which will make it easier to run slither locally without the noise

File: packages/contracts-bedrock/package.json
Comment (tynes) [inline packages/contracts-bedrock/package.json]: The following 1 liner can replace the tempfile approach

```
cat <<< $(jq 'QUERY' sample.json) > sample.json
```

Comment (mds1) [reply packages/contracts-bedrock/package.json]: Hmm, seems to not outputting correctly and just places the full contents on 1 line. Can debug it more tomorrow

Comment (mds1) [reply packages/contracts-bedrock/package.json]: Decided we can start by trying github code scanning integration approach to minimize DB management and simplify the triage process

File: packages/contracts-bedrock/.gitignore
Comment (mds1) [inline packages/contracts-bedrock/.gitignore:1]: _(arbitrary comment location to start a thread)_

@clabby Should this be `i > 0` or can that `i >= 0` condition be removed? https://github.com/ethereum-optimism/optimism/security/code-scanning/36

File: packages/contracts-bedrock/README.md
Comment (mds1) [inline packages/contracts-bedrock/README.md:97]: _(arbitrary comment location to start a thread)_

This finding seems valid, should we remove the payable modifier? Or is this changing anyway with fault proof work? https://github.com/ethereum-optimism/optimism/security/code-scanning/26

File: .github/workflows/slither.yml
Comment (mds1) [inline .github/workflows/slither.yml:33]: `slither-version` will be removed once this `--triage-database` feature is incorporated into a released version of slither

File: .github/workflows/slither.yml
Comment (mds1) [inline .github/workflows/slither.yml:24]: This will use the latest foundry nightly instead of our pinned version, which I think is ok for now since it shouldn't affect slither results. Can switch to ci builder image in a future PR

**Review by tynes — APPROVED** [review]

**Comment by tynes** [comment]
regarding the `L2OutputOracle`, it is payable to save a little bit of gas when submitting outputs. The architecture also changed but it would have been forward compatible with sending bonds

File: packages/contracts-bedrock/README.md
Comment (mds1) [reply packages/contracts-bedrock/README.md:97]: Thanks @tynes for the answer in https://github.com/ethereum-optimism/optimism/pull/9405#issuecomment-1934610890, I've dismissed this alert

File: packages/contracts-bedrock/.gitignore
Comment (clabby) [reply packages/contracts-bedrock/.gitignore:1]: No this looks correct. See the `unchecked` decrementation.

Comment (mds1) [reply packages/contracts-bedrock/.gitignore:1]: Thanks! Just dismissed the alert

**Pull request merged**
